app/models/manageiq/providers/openstack/cloud_manager/cloud_volume_snapshot.rb

```diff
@@ -1,2 +1,138 @@
+    ext_management_system = cloud_volume.try(:ext_management_system)
+    raise ArgumentError, _("ext_management_system cannot be nil") if ext_management_system.nil?
+
+    cloud_tenant = cloud_volume.cloud_tenant
+    snapshot = nil
+    options[:volume_id] = cloud_volume.ems_ref
+    ext_management_system.with_provider_connection(connection_options(cloud_tenant)) do |service|
+      snapshot = service.snapshots.create(options)
+    end
+
+    create(
+      :name                  => snapshot.name,
+      :description           => snapshot.description,
+      :ems_ref               => snapshot.id,
+      :status                => snapshot.status,
+      :cloud_volume          => cloud_volume,
+      :cloud_tenant          => cloud_tenant,
+      :ext_management_system => ext_management_system,
+    )
+  rescue => e
+    _log.error "snapshot=[#{options[:name]}], error: #{e}"
+    raise MiqException::MiqVolumeSnapshotCreateError, e.to_s, e.backtrace
+  end
+
+  def update_snapshot_queue(userid = "system", options = {})
+    task_opts = {
+      :action => "updating volume snapshot #{inspect} in #{ext_management_system.inspect} with #{options.inspect}",
+      :userid => userid
+    }
+
+    queue_opts = {
+      :class_name  => self.class.name,
+      :instance_id => id,
+      :method_name => 'update_snapshot',
+      :priority    => MiqQueue::HIGH_PRIORITY,
+      :role        => 'ems_operations',
+      :zone        => my_zone,
+      :args        => [options]
+    }
+
+    MiqTask.generic_action_with_callback(task_opts, queue_opts)
+  end
+
+  def update_snapshot(options = {})
+    with_provider_object do |snapshot|
+      if snapshot
+        snapshot.update(options)
+      else
+        raise MiqException::MiqVolumeSnapshotUpdateError("snapshot does not exist")
+      end
+    end
+  rescue => e
+    _log.error "snapshot=[#{name}], error: #{e}"
+    raise MiqException::MiqVolumeSnapshotUpdateError, e.to_s, e.backtrace
+  end
+
+  def delete_snapshot_queue(userid = "system", _options = {})
+    task_opts = {
+      :action => "deleting volume snapshot #{inspect} in #{ext_management_system.inspect}",
+      :userid => userid
+    }
+
+    queue_opts = {
+      :class_name  => self.class.name,
+      :instance_id => id,
+      :method_name => 'delete_snapshot',
+      :priority    => MiqQueue::HIGH_PRIORITY,
+      :role        => 'ems_operations',
+      :zone        => my_zone,
+      :args        => []
+    }
+
+    MiqTask.generic_action_with_callback(task_opts, queue_opts)
+  end
+
+  def delete_snapshot(_options = {})
+    with_provider_object do |snapshot|
+      if snapshot
+        snapshot.destroy
+      else
+        _log.warn "snapshot=[#{name}] already deleted"
+      end
+    end
+  rescue => e
+    _log.error "snapshot=[#{name}], error: #{e}"
+    raise MiqException::MiqVolumeSnapshotDeleteError, e.to_s, e.backtrace
+  end
+
+  private
+
+  def self.connection_options(cloud_tenant = nil)
```

Comment:
@petrblaho was the intention make this class method private?

Because it doesn't work in this way.  You need to use `private_class_method`.  See codeclimate screenshot.  Note, we prefer `private_class_method` over `class << self` block notation.


![image](https://cloud.githubusercontent.com/assets/19339/19650398/28cbf4fe-99d7-11e6-9bf3-a862799f915d.png)